### ISSUE TITLE
WARN...java.io.IOException: Connection reset by peer

### ISSUE THREAD

**Issue by mobilesoftsmith** [relayrides/pushy#272]
Hi,

First of all, thanks for your great work.

I have a push server running as a service on Linux using 0.6.1. I find this warning showing up after running 1 or 2 hours. The notification can still be sent out. It is just annoying.

```
[nioEventLoopGroup-4-1] WARN com.relayrides.pushy.apns.ApnsClientHandler - APNs client pipeline caught an exception.
java.io.IOException: Connection reset by peer
at sun.nio.ch.FileDispatcherImpl.read0(Native Method)
at sun.nio.ch.SocketDispatcher.read(SocketDispatcher.java:39)
at sun.nio.ch.IOUtil.readIntoNativeBuffer(IOUtil.java:223)
at sun.nio.ch.IOUtil.read(IOUtil.java:192)
at sun.nio.ch.SocketChannelImpl.read(SocketChannelImpl.java:379)
at io.netty.buffer.PooledUnsafeDirectByteBuf.setBytes(PooledUnsafeDirectByteBuf.java:288)
at io.netty.buffer.AbstractByteBuf.writeBytes(AbstractByteBuf.java:1054)
at io.netty.buffer.WrappedByteBuf.writeBytes(WrappedByteBuf.java:783)
at io.netty.channel.socket.nio.NioSocketChannel.doReadBytes(NioSocketChannel.java:245)
at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:106)
at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:527)
at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:484)
at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:398)
at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:370)
at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:742)
at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:145)
at java.lang.Thread.run(Thread.java:745)
```

And separately, this message appears frequently also:

```
[nioEventLoopGroup-8-1] INFO com.relayrides.pushy.apns.ApnsClient - Connection to api.development.push.apple.com/17.172.238.203:443 restored.
```

Is this normal?

**Comment by jchambers** [comment]
> Is this normal?

Well, we're restoring the connection automatically, so I'd say that's "normal" behavior in the face of connection closure. That the connection is closing at all is a little surprising, though. If the server initiates connection closure, you should see a line like this somewhere in the logs (assuming you're logging at the `INFO` level):

> Received GOAWAY from APNs server

Are you seeing any of those? Also, how frequently are you sending notifications?

**Comment by jchambers** [comment]
…and, to clarify, do you see that exception when you actually send a notification, or does it happen spontaneously?

**Comment by mobilesoftsmith** [comment]
I used to get "Received GOAWAY from APNs server" error. It appears when I switch to production APNs server. I think it may be because my app is in debug mode. I will publish my app with distribution code signing id and do more testings with production APNs server to see if I get this error again. By design, my push server should send massive notification per second. But right now, I just send a few for testing. 

To your second question, the exception just happens spontaneously. My push server has other issues, so I leave it running over night without sending any notification. This exception still appears. Hope this information is helpful.

**Comment by jchambers** [comment]
> I used to get "Received GOAWAY from APNs server" error.

Used to? But not any more? Forgive me if I'm being a stickler on this point, but I think it'll help us understand what, exactly, is happening here. When you did receive those messages, what was the reason given by the server for sending the `GOAWAY` frame?

> I leave it running over night without sending any notification.

To confirm, you're using Pushy 0.6.1, not 0.6.0? 0.6.1 added support for sending `PING` frames when the connection is idle; if the connection is still spontaneously closing, that will be an important detail.

**Comment by mobilesoftsmith** [comment]
I received the "GOAWAY" error when I switched to production server using my product.p12 file. But I found there was some problem in my Apple provision profile and certificates. I recreated everything and "GOAWAY" error was gone. Now I received "BadDeviceToken" error when I use production APNs (I can send through sandbox APNs without problem, thanks for your work). I believe it is caused by the develop code sign credential I used in my app. I have submitted my app to Apple Store for reviewing. Hopefully, the production code sign credential will resolve this issue.

I am using 0.6.1. It looks like this version will recreate the connection if this error happens. I forgot to tell you that I was running my server in a VM on my laptop. I think the internet was disconnected after a while when I went to bed. I didn't see this error on my server again so far. Thanks,
